man/omorfi-lookup.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.46.6.
.TH OMORFI "1" "June 2018" "OMORFI" "User Commands"
.SH NAME
omorfi-lookup \- Toy example for omorfi C++ API
.SH SYNOPSIS
.B omorfi-lookup
[\fI\,OPTIONS\/\fR...] \fI\,FSAFILE\/\fR
.SH DESCRIPTION
Lookup tokens using omorfi
.TP
\fB\-h\fR, \fB\-\-help\fR
print this help
.TP
\fB\-V\fR, \fB\-\-version\fR
show version info
.SH BUGS
This is just a toy demo example written in few minutes.
.SH "SEE ALSO"
omorfi-analyse-text(1), omorfi-disambiguate-text(1)


NEWS
# NEWS

This file lists noteworthy changes between releases, for full list of changes,
see git log and then `ChangeLog.old`.

## Significant changes in 20180511

* Universal dependencies version 2.2 is now used as target
* At least 226 new words: 239 additions and 13 deletions in lexeme database
* Most changes are in development infra, so not visible to end users...
* Started rewriting CG from the scratch
* The APIs for programming language deprecate load(filename) and load(dir) forms
  of filename guessing functions in favour of forthcoming loadAnalyser(file),
  loadLemmatiser(file), loadUDPipe(file) etc. etc. functions
* Working towards more general tokenise-analyse-disambiguate pipelines maybe,
  or just refactoring
* lots more automated tests -> lots less human errors
* By popular request: there are two analysers now, one with small dictionary
  and one with full, use the smaller one when you do not want to see birds or
  languages or tribes analysed. The smaller one replaces the old default, but
  the new tools will require you to select one explicitly anyways
* fixes and workarounds: java and c++ can now be disabled partially or totally
* adopted SG0 as possible verb form analysis from UD data
* The end users are now provided with bash-scripts wrappers for all
  functionalities, whereas the typically python versions allow more control
  of parametres


## Significant changes in 20170515

* Universal Dependencies version 2 is now used, still mainly lemma, UPOS,
  features fields are analysed
* At least 2,336 new words (based on diffstat: 38886 additions, 3655 deletions)
* Preliminary support for various guessing models: python-based, finite-state
  and UDPipe. This means that it is possibly to get analyses for all tokens,
  albeit quality of guesses varies.
* A minimal C++ library version has been made to match java and python bindings.
  C++-11 and libhfst are required.
* The dix version can now be compiled with lttoolbox with a lot of memory
* A restricted "gold" dictionary mode has been added. This is good for both end
  users with limited memory and end users who require higher quality lexemes
  (i.e., only research institute approved, no wiktionary words or other weird
   stuffs)
* Documentations and automatic testing much reworked with the new modern toys
  from github: travis-ci, jekyll
* Started weeding the ADP/ADV jungle...
* Fixed a horrible bug in the corpus coverage testing that terribly
  under-estimated our coverage for corpora where hapax legomena etc. were
  ignored
* Lot of documentation has been semi-automated, therefore many changes can be
  viewed at the new gh-pages site: https://flammie.github.io/omorfi/

## Significant changes in 20161115

* Started drafting more blacklists and *known good* lexemes subsets for people
  who struggle with rare words and productive compounding, derivation
* Updated to Universal Dependencies version 1.4
* A lot of new derivations by the way
* Preliminary guessers
* More loopy guessery things for punctuation and digit combos
* Minor fixes to UD feature sorting
* Homonym numbers used in some applications
* Added timeouts where downstream tools support them, so tools don't seem like
  they are freezing at random
* moved old documentations to github-pages
* added preliminary hfst-pmatch-based tokeniser

## Significant changes in 20160515

* Universal Dependencies for Finnish is the new standard format we now follow:
    * POS is now UPOS and classes were changed accordingly (new classes: AUX,
      PROPN, DET, CONJ, SCONJ, PUNCT, SYM, and VERB, NOUN, ADP, ADV as before)
    * other features mostly match the feature field in UD documentation
    * release cycle aims to be same six month cycle as with UD
    * the automatic tests verify compatibility with UD; 92 % of lemmas, primary
      POS tags and morphological features are the same as Finnish UD corpus,
      75 % same as Finnish FTB UD corpus
    * analyser for reading and writing CONLL-U format
* tokenisation as script and more hacks to token stripping in corner cases
* continuous integration with travis-ci, currently only testing basic script
  programming conventions
* added a lot of high coverage words and forms by hand
* by popular request, some of the words can now be blacklisted, when you don't
  want that guy named Mutta to ambiguate your conjunction analyses or the odd
  new guinean bird to clash with some common verb
* the "database" is now only keyed on lemma + homonym number; paradigm is extra
  information like anything else
* a lot of work on morphological segmentation towards statistical machine
  translation; check proceedings of WMT shared tasks 2015 and 2016 to see why
* started refactoring some python code into classes

## Significant changes in 20150904

* allomorphy can be tagged again to distinguish e.g. *-iden* and *-itten* when
  generating
* FinnTreeBank-1 format provided by Miikka Silfverberg is available but not
  built by default since it lacks a test set
* lexicalised inflections can have separate tag, e.g. *kännissä* can be lexical
  inessive distinguished from regular inessive
* preliminary VISL CG-3 support, with original grammar by Fred Karlsson;
  convenience bash scripts available for disambiguated parsing
* preliminary support for conllu and conllx analysis formats
* paradigm categorisation is now verified by regular expressions
* lots of paradigm fixes and some added words

## Significant changes in 20150326

* speed is up to >20,000 tokens per second from ~500
* coverages are up to:
  europarl (99 %) gutenberg (97 %), JRC Acquis (94 %) and fiwiki (93 %)
* moses factored model format supported
* segmentation supported
* Java API
* Python hacks packaged to API and module
* Rest of hand-written Xerox legacy data removed; all is script-generated
* github migration since google code is EOL'd
* file naming for automata changed to include omorfi prefix for all file
  names in case they are distributed separately.

## Significant changes in 20141014


src/cpp/omorfi-lookup.cc
/**
 * @file omorfi-lookup.cc
 *
 * @brief Example demo of omorfi C++ API usage.
 */

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif


#include <getopt.h>

#include "omorfi.hh"

/** Print standard usage message.
 *
 * @param program_name name of the program
 */
void
print_usage(const char* program_name) {
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    // Usage line
    printf("Usage: %s [OPTIONS...] FSAFILE\n"
           "Lookup tokens using omorfi\n"
        "\n", program_name);
    printf("  -h, --help      print this help\n");
    printf("  -V, --version   show version info\n");
    printf("\n");
}

/** Command-line interface to omorfi lookup.
 *
 * @param argc  number of arguments
 * @param argv  arguments in array
 * @return 0 if program terminated succesfully, 1 otherwise
 */
int main(int argc, char** argv) {
    while (true) {
        static const struct option long_options[] = {
              {"help", no_argument, 0, 'h'},
              {"version", no_argument, 0, 'V'},
              {0,0,0,0}
        };
        int option_index = 0;
        int c = getopt_long(argc, argv, "hV", long_options, &option_index);
        if (-1 == c) {
            break;
        }
        switch (c) {
          case 'h':
            print_usage(argv[0]);
            exit(0);
            break;
          case 'V':
            printf("%s\n", PACKAGE_STRING);
            exit(0);
            break;
          case '?':
          case ':':
          default:
            printf("Some error in getopt handling, maybe -%c\n", optopt);
            break;
        }
    }
    if ((argc - optind) == 1) {
        std::string fsafilename(argv[optind]);
        printf("Opening %s for omorfi\n", fsafilename.c_str());
        omorfi::Omorfi omorfi;
        omorfi.loadAnalyser(fsafilename);
        printf("Reading one token per line:\n");
        char* line = nullptr;
        size_t len = 0;
        ssize_t read;
        while ((read = getline(&line, &len, stdin)) != -1) {
            std::string token(line);
            // getline includes \n; should never have more than one \n
            if (token.find("\n")) {
                token = token.substr(0, token.rfind("\n"));
            }
            std::vector<std::string> analyses = omorfi.analyse(token);
            printf("→ %s:\n", token.c_str());
            for (std::string analysis : analyses) {
                printf("← %s\n", analysis.c_str());
            }
        }
    } else {
        printf("Exactly one FSAFILE argument is needed\n");
        exit(1);
    }
}



